---
title: "Formatting a SpatialExperiment Object for SEraster"
output: html_document
vignette: >
  %\VignetteIndexEntry{formatting-SpatialExperiment-for-SEraster}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width=12, fig.height=8
)
```

# Formatting a SpatialExperiment Object for SEraster

For this tutorial, we will format a preprocessed MERFISH dataset of the mouse preotic area (POA) into a `SpatialExperiment` so that it can be rasterized with `SEraster`. The output of this tutorial is the same as the `merfish_mousePOA` dataset in the package.

In the original work, [Moffitt J. and Bambah-Mukku D. et al. (2018), "Molecular, spatial, and functional single-cell profiling of the hypothalamic preoptic region", *Science Advances*](https://www.science.org/doi/10.1126/science.aau5324), authors collected spatial transcriptomics datasets of mouse POA regions for various sexes, behavioral conditions, and bregma sections. The full dataset can be downloaded from [Dryad](https://datadryad.org/stash/dataset/doi:10.5061/dryad.8t8s248). We only use the dataset for a bregma -0.29 slice from a female naive animal (Animal ID = 1).

## Load libraries

```{r}
library(SpatialExperiment)
library(Matrix)
library(ggplot2)
library(gridExtra)
```

## Load dataset

```{r}
data <- read.csv('~/Downloads/Moffitt_and_Bambah-Mukku_et_al_merfish_all_cells.csv')
```

Subset the dataset to specific female, behavioral condition, and bregma section.

```{r}
animal <- 1
sex <- "Female"
behavior <- "Naive"
bregma <- "-0.29"
data_sub <- data[(data$Animal_ID == animal & data$Animal_sex == sex & data$Behavior == behavior & data$Bregma == bregma),]
dim(data_sub)
```

## Process dataset

The subsetted has cells. For rasterizing gene expression with `SEraster`, the input `SpatialExperiment` needs to have a genes-by-cells matrix (either dense or sparse matrix) in the `assay` slot and a spatial coordinates matrix in the `spatialCoords` slot. For rasterizing cell-type labels with `SEraster`, in addition to the spatial coordinates matrix in the `spatialCoords` splot, the input `SpatialExperiment` also need to have a data frame with a column for cell-type labels in the `colData` slot. Thus, here, we extract genes-by-cells matrix, spatial coordinates matrix, and data frame with cell-type labels.

```{r}
## genes-by-cells matrix
# extract the genes-by-cells matrix as a sparse matrix (dgCMatrix)
mat <- as(t(data_sub[,10:ncol(data_sub)]), "CsparseMatrix")

# remove blank genes used for quality control
blanks <- rownames(mat)[grepl("Blank", rownames(mat))]
mat <- mat[setdiff(rownames(mat),blanks),]

## spatial coordinates matrix
# extract the spatial coordinates
pos <- data_sub[,c("Centroid_X", "Centroid_Y")]
colnames(pos) <- c("x","y")

# make x,y coordinates positive
pos[,1] <- pos[,1] - min(pos[,1])
pos[,2] <- pos[,2] - min(pos[,2])

## cell-type labels
# extract the data frame with cell-type labels
meta <- data_sub[,c("Bregma", "Cell_class", "Neuron_cluster_ID")]
colnames(meta) <- c("bregma", "celltype", "neurontype")

## standardize cell IDs for the extracted objects
colnames(mat) <- rownames(pos) <- rownames(meta) <- data_sub$Cell_ID
```

Remove genes and cells with NaN values.

```{r}
## filter genes with NaN values
bad_genes <- names(which(rowSums(is.nan(mat)) > 0)) 
mat <- mat[setdiff(rownames(mat),bad_genes),]

## filter cells with NaN values
bad_cells <- names(which(colSums(is.nan(mat)) > 0))
mat <- mat[,setdiff(colnames(mat),bad_cells)]
pos <- pos[setdiff(rownames(pos),bad_cells),]
meta <- meta[setdiff(rownames(pos),bad_cells),]
```

Plot total gene expression at single-cell resolution for verification.

```{r}
df_plt <- data.frame(pos, total_gexp = colSums(mat))

ggplot(df_plt, aes(x = x, y = y, color = total_gexp)) +
  coord_fixed() +
  geom_point(size = 1.5, stroke = 0) +
  scale_color_viridis_c(name = "total gene expression") +
  theme_bw() +
  theme(panel.grid = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())
```

Plot cell-type labels at single-cell resolution for verification.

```{r}
df_plt <- data.frame(pos, celltype = meta$celltype)

ggplot(df_plt, aes(x = x, y = y, color = celltype)) +
  coord_fixed() +
  geom_point(size = 1.5, stroke = 0) +
  theme_bw() +
  theme(panel.grid = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())
```

## Format SpatialExperiment object

Format genes-by-cells matrix, spatial coordinates matrix, and data frame with cell-type labels into a `SpatialExperiment` object. Here, the genes-by-cells matrix is named as "volnorm" because the loaded gene expression was already normalized by cell volume and scaled by 1000.

```{r}
spe <- SpatialExperiment::SpatialExperiment(
  assays = list(volnorm = mat),
  spatialCoords = as.matrix(pos),
  colData = meta
)
```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>