---
title: "SEraster with Xenium Lymph Node"
author: "Jean Fan"
date: "3/10/2024"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Xenium output bundle frome: https://www.10xgenomics.com/datasets/human-lymph-node-preview-data-xenium-human-multi-tissue-and-cancer-panel-1-standard

```{r}
dir <- '~/OneDrive - Johns Hopkins/Data_Public/xenium_data/human-lymph-node-preview-data-xenium-human-multi-tissue-and-cancer-panel-1-standard/Xenium_V1_hLymphNode_nondiseased_section_outs/'
```

Read in cell positions stored in `cells.parquet` using `sparklyr`.

```{r}
## read in cell positions

## https://therinspark.com/starting.html
#install.packages("sparklyr")
library(sparklyr)
#spark_available_versions()
#spark_install("3.5")
spark_session <- spark_connect(master = "local", version="3.5")
parquet_data_spark <- spark_read_parquet(spark_session, name = "parquet_data", path = paste0(dir, 'cells.parquet'))
pos.info <- collect(parquet_data_spark)
spark_disconnect(spark_session)

head(pos.info)
```

Read in the sparse gene expression matrix stored in `cell_feature_matrix.h5` using `rhdf5`.

```{r}
## BiocManager::install("rhdf5")
library(rhdf5)
h5file <- paste0(dir, "cell_feature_matrix.h5")
h5ls(h5file)

h5 <- h5read(h5file, "matrix")
barcodes <- as.character(h5read(h5file, "matrix/barcodes"))

library(Matrix)
counts <- sparseMatrix(
    dims = h5$shape,
    i = as.numeric(h5$indices),
    p = as.numeric(h5$indptr),
    x = as.numeric(h5$data),
    index1 = FALSE
  )
colnames(counts) <- barcodes
rownames(counts) <- as.data.frame(h5[["features"]])$name

head(counts)
```

Visualize cell metadata using `ggplot2`.

```{r}
library(ggplot2)
ggplot(data.frame(pos.info)) + 
  geom_point(aes(x = x_centroid, y = y_centroid, col=cell_area), 
             size=0.01, alpha=0.1) +
  scale_color_gradient(low = 'lightgrey', high='red') + 
  theme_minimal()
```

Create `SpatialExperiments` object.

```{r}
library(SpatialExperiment)
pos <- as.matrix(cbind(x=pos.info$x_centroid, y=pos.info$y_centroid))
rownames(pos) <- pos.info$cell_id
se <- SpatialExperiment(
  assays = list(counts = counts),
  spatialCoords = pos[colnames(counts),]
)
```

Create `SEraster`

```{r}
res <- 100
rastGexp <- SEraster::rasterizeGeneExpression(se, assay_name="counts", resolution = res, square=FALSE)
```
Identify spatially variable genes using Moran's I on rasterized gene expression output

```{r}
rastpos <- spatialCoords(rastGexp)

library(MERINGUE)
w <- getSpatialNeighbors(rastpos, filterDist = res)
plotNetwork(rastpos, w, cex=0.01)
```
```{r}
# Identify sigificantly spatially auto-correlated genes
I <- getSpatialPatterns(assay(rastGexp, 'pixelval'), w)
```

```{r}
# plot
g <- "MS4A1"
print(I[g,])
SEraster::plotRaster(rastGexp, feature_name = g, name = g)
```
```{r}
# plot
g <- "BLANK_0499"
print(I[g,])
SEraster::plotRaster(rastGexp, feature_name = g, name = g)
```
