<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SEraster is a rasterization preprocessing framework that aggregates  
    cellular information into spatial pixels to reduce resource requirements for  
    spatial omics data analysis. SEraster reduces the number of spatial points in  
    spatial omics datasets for downstream analysis through a process of rasterization  
    where single cells’ gene expression or cell-type labels are aggregated into  
    equally sized pixels based on a user-defined resolution. SEraster is built on  
    an R/Bioconductor S4 class called SpatialExperiment. SEraster can be incorporated  
    with other packages to conduct downstream analyses for spatial omics datasets,  
    such as detecting spatially variable genes.">
<title>Rasterization Preprocessing Framework for Scalable Spatial Omics Data Analysis • SEraster</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Rasterization Preprocessing Framework for Scalable Spatial Omics Data Analysis">
<meta property="og:description" content="SEraster is a rasterization preprocessing framework that aggregates  
    cellular information into spatial pixels to reduce resource requirements for  
    spatial omics data analysis. SEraster reduces the number of spatial points in  
    spatial omics datasets for downstream analysis through a process of rasterization  
    where single cells’ gene expression or cell-type labels are aggregated into  
    equally sized pixels based on a user-defined resolution. SEraster is built on  
    an R/Bioconductor S4 class called SpatialExperiment. SEraster can be incorporated  
    with other packages to conduct downstream analyses for spatial omics datasets,  
    such as detecting spatially variable genes.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">SEraster</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/getting-started-with-SEraster.html">Getting Started With SEraster</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JEFworks-Lab/SEraster/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="spatial-experiments-raster-seraster">Spatial Experiments raster (SEraster)<a class="anchor" aria-label="anchor" href="#spatial-experiments-raster-seraster"></a>
</h1></div>
<p><code>SEraster</code> is a rasterization preprocessing framework that aggregates cellular information into spatial pixels to reduce resource requirements for spatial omics data analysis.</p>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/seraster_logo_hex.png?raw=true" height="200"></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>SEraster</code> reduces the number of spatial points in spatial omics datasets for downstream analysis through a process of rasterization where single cells’ gene expression or cell-type labels are aggregated into equally sized pixels based on a user-defined <code>resolution</code>. Here, we refer to a particular <code>resolution</code> of rasterization by the side length of the pixel such that finer <code>resolution</code> indicates smaller pixel size and coarser <code>resolution</code> indicates larger pixel size.</p>
<p align="center">
<img src="https://github.com/JEFworks-Lab/SEraster/blob/main/images/overview.png?raw=true" height="600"></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install <code>SEraster</code>, we recommend using <code>remotes</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'JEFworks-Lab/SEraster'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
</div>
<div class="section level2">
<h2 id="input-data-format">Input data format<a class="anchor" aria-label="anchor" href="#input-data-format"></a>
</h2>
<p>In the examples below, we assume the input data is provided as a <code>SpatialExperiment</code> Bioconductor object. Please refer to the <a href="https://bioconductor.org/packages/SpatialExperiment" class="external-link">SpatialExperiment</a> package and the <code>merfish_mousePOA</code> dataset in the package to see how you would format your data into a <code>SpatialExperiment</code> object.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>A short example workflow is shown below.</p>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SEraster</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-example-dataset">Load example dataset<a class="anchor" aria-label="anchor" href="#load-example-dataset"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"merfish_mousePOA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the dimension of the genes-by-cells matrix at single-cell resolution</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>[<span class="dv">1</span>]  <span class="dv">155</span> <span class="dv">6509</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># check the number of cell-types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>[<span class="dv">1</span>]  <span class="dv">16</span></span></code></pre></div>
<p>This MERFISH mouse preoptic area dataset contains 6,509 cells and 16 cell-types.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot at single-cell resolution</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span>, celltype <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, col <span class="op">=</span> <span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"x (μm)"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"y (μm)"</span>,</span>
<span>       col <span class="op">=</span> <span class="st">"Cell-types"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks-Lab/SEraster/blob/main/images/singlecell_celltypes.png?raw=true" height="550"></p>
</div>
<div class="section level3">
<h3 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h3>
<div class="section level4">
<h4 id="rasterize-gene-expression">Rasterize gene expression<a class="anchor" aria-label="anchor" href="#rasterize-gene-expression"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rastGexp</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the dimension of the genes-by-cells matrix after rasterizing gene expression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>[<span class="dv">1</span>]  <span class="dv">155</span> <span class="dv">1301</span></span></code></pre></div>
<p>As you can see, SEraster aggregated 6,509 single cells into 1,301 pixels.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot total rasterized gene expression</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, name <span class="op">=</span> <span class="st">"Total rasterized gene expression"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_gexp_total.png?raw=true" height="400"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot specific gene</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, feature_name <span class="op">=</span> <span class="st">"Esr1"</span>, name <span class="op">=</span> <span class="st">"Esr1"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_gexp_esr1.png?raw=true" height="400"></p>
</div>
<div class="section level4">
<h4 id="rasterize-cell-type-labels">Rasterize cell-type labels<a class="anchor" aria-label="anchor" href="#rasterize-cell-type-labels"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rastCt</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/rasterizeCellType.html">rasterizeCellType</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, col_name <span class="op">=</span> <span class="st">"celltype"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the dimension of the cell-types-by-cells matrix after rasterizing cell-type labels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>[<span class="dv">1</span>]  <span class="dv">16</span> <span class="dv">1301</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot total cell counts</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, name <span class="op">=</span> <span class="st">"cell counts"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_ct_total.png?raw=true" height="400"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot specific cell-type</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, feature_name <span class="op">=</span> <span class="st">"Inhibitory"</span>, name <span class="op">=</span> <span class="st">"Inhibitory neuron counts"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_ct_inhibitory.png?raw=true" height="400"></p>
</div>
</div>
<div class="section level3">
<h3 id="sample-downstream-analysis">Sample Downstream Analysis<a class="anchor" aria-label="anchor" href="#sample-downstream-analysis"></a>
</h3>
<p><code>SEraster</code> returns rasteriezd gene-expression and cell-type information as <code>SpatialExperiment</code> objects that can be integrated with other existing downstream analysis tools. We demonstrate below spatial variable gene (SVG) and cell-type cooccurrence analyses as examples of such potential downstream analysis.</p>
<div class="section level4">
<h4 id="spatial-variable-gene-svg-analysis">Spatial variable gene (SVG) analysis<a class="anchor" aria-label="anchor" href="#spatial-variable-gene-svg-analysis"></a>
</h4>
<p>Here, we use a previously developed tool called <code>nnSVG</code>. Please refer to <a href="https://bioconductor.org/packages/nnSVG" class="external-link">nnSVG</a> for more details about the package. We can directly input rasterized gene expression <code>SpatialExperiment</code> object from <code>SEraster</code> into <code>nnSVG</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nnSVG</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run nnSVG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">rastGexp</span> <span class="op">&lt;-</span> <span class="fu">nnSVG</span><span class="op">(</span><span class="va">rastGexp</span>, assay_name <span class="op">=</span> <span class="st">"pixelval"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of significant SVGs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE</span></span>
<span><span class="co">##    17   138</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot rasterized gene expression of top-ranked SVG</span></span>
<span><span class="va">top_svg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">$</span><span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">top_svg_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">top_svg</span><span class="op">]</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, feature_name <span class="op">=</span> <span class="va">top_svg_name</span>, name <span class="op">=</span> <span class="va">top_svg_name</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_gexp_top_svg.png?raw=true" height="400"></p>
<p>We can also perform cell-type specific SVG analysis by subsetting the dataset prior to applying SEraster.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset data</span></span>
<span><span class="va">ct_interest</span> <span class="op">&lt;-</span> <span class="st">"Excitatory"</span></span>
<span><span class="va">spe_sub</span> <span class="op">&lt;-</span> <span class="va">merfish_mousePOA</span><span class="op">[</span>,<span class="va">merfish_mousePOA</span><span class="op">$</span><span class="va">celltype</span> <span class="op">==</span> <span class="va">ct_interest</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># run SEraster</span></span>
<span><span class="va">rastGexp_sub</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">spe_sub</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run nnSVG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">rastGexp_sub</span> <span class="op">&lt;-</span> <span class="fu">nnSVG</span><span class="op">(</span><span class="va">rastGexp_sub</span>, assay_name <span class="op">=</span> <span class="st">"pixelval"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of significant SVGs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp_sub</span><span class="op">)</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##    45   110</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot rasterized gene expression of top-ranked SVG</span></span>
<span><span class="va">top_svg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp_sub</span><span class="op">)</span><span class="op">$</span><span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">top_svg_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp_sub</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">top_svg</span><span class="op">]</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp_sub</span>, feature_name <span class="op">=</span> <span class="va">top_svg_name</span>, name <span class="op">=</span> <span class="va">top_svg_name</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_gexp_sub_top_svg.png?raw=true" height="400"></p>
</div>
<div class="section level4">
<h4 id="cell-type-cooccurrence-analysis">Cell-type cooccurrence analysis<a class="anchor" aria-label="anchor" href="#cell-type-cooccurrence-analysis"></a>
</h4>
<p>Rasterized cell-type labels can be used to analyze pair-wise cell-type cooccurrence. To do so, we binarize the rasterized cell-type labels using a relative enrichment metric and a previously developed tool called <code>CooccurrenceAffinity</code>. Please refer to our paper for more details about the methodology and <a href="https://CRAN.R-project.org/package=CooccurrenceAffinity" class="external-link">CooccurrenceAffinity</a> for more details about the package.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kpmainali/CooccurrenceAffinity" class="external-link">CooccurrenceAffinity</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract cell-type labels</span></span>
<span><span class="va">ct_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute relative enrichment (RE) metric</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">rastCt</span>, <span class="st">"pixelval"</span><span class="op">)</span></span>
<span><span class="va">mat_re</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rastCt</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ct_label</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">mat</span><span class="op">[</span><span class="va">ct_label</span>,<span class="op">]</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mat</span><span class="op">[</span><span class="va">ct_label</span>,<span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat_re</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># binarize</span></span>
<span><span class="va">mat_bin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">mat_re</span> <span class="op">&gt;=</span> <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add RE and binarized layers to SpatialExperiment object</span></span>
<span><span class="fu">assays</span><span class="op">(</span><span class="va">rastCt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pixelval <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">rastCt</span>, <span class="st">"pixelval"</span><span class="op">)</span>, re <span class="op">=</span> <span class="va">mat_re</span>, bin <span class="op">=</span> <span class="va">mat_bin</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_interest</span> <span class="op">&lt;-</span> <span class="st">"Ependymal"</span></span>
<span></span>
<span><span class="co"># plot pixel value for a cell-type of interest</span></span>
<span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, assay_name <span class="op">=</span> <span class="st">"pixelval"</span>, feature_name <span class="op">=</span> <span class="va">ct_interest</span>, name <span class="op">=</span> <span class="st">"cell-type counts"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_ct_sub_total.png?raw=true" height="400"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot RE value for a cell-type of interest</span></span>
<span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, assay_name <span class="op">=</span> <span class="st">"re"</span>, feature_name <span class="op">=</span> <span class="va">ct_interest</span>, name <span class="op">=</span> <span class="st">"RE"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_ct_sub_re.png?raw=true" height="400"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot binarized value for a cell-type of interest</span></span>
<span><span class="fu"><a href="reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, assay_name <span class="op">=</span> <span class="st">"bin"</span>, feature_name <span class="op">=</span> <span class="va">ct_interest</span>, factor_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"binarized"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/rasterized_ct_sub_bin.png?raw=true" height="400"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run CooccurrenceAffinity</span></span>
<span><span class="va">ct_coocc</span> <span class="op">&lt;-</span> <span class="fu">CooccurrenceAffinity</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/CooccurrenceAffinity/man/affinity.html" class="external-link">affinity</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mat_bin</span>, row.or.col <span class="op">=</span> <span class="st">"row"</span>, squarematrix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot maximum likelihood estimates of affinity metric (alpha MLE)</span></span>
<span><span class="fu">CooccurrenceAffinity</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/CooccurrenceAffinity/man/plotgg.html" class="external-link">plotgg</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ct_coocc</span>, variable <span class="op">=</span> <span class="st">"alpha_mle"</span>, legendlimit <span class="op">=</span> <span class="st">"datarange"</span><span class="op">)</span></span></code></pre></div>
<p align="center">
<img src="https://github.com/JEFworks/SEraster/blob/main/images/coocc_heatmap.png?raw=true" height="500"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Our preprint describing <code>SEraster</code> is available on bioRxiv:</p>
<p><a href="https://doi.org/10.1101/2024.02.01.578436" class="external-link">Aihara G. et al. (2024), “SEraster: a rasterization preprocessing framework for scalable spatial omics data analysis”, bioRxiv</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/JEFworks-Lab/SEraster/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/JEFworks-Lab/SEraster/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SEraster</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Gohta Aihara <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2492-9610" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lyla Atta <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6113-0082" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jean Fan <br><small class="roles"> Author, reviewer </small> <a href="https://orcid.org/0000-0002-0212-5451" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gohta Aihara, Lyla Atta, Jean Fan.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
