<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Rasterization Preprocessing Framework for Scalable Spatial Omics Data Analysis • SEraster</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Rasterization Preprocessing Framework for Scalable Spatial Omics Data Analysis">
<meta name="description" content="SEraster is a rasterization preprocessing framework that aggregates cellular information into spatial pixels to reduce resource requirements for spatial omics data analysis. SEraster reduces the number of spatial points in spatial omics datasets for downstream analysis through a process of rasterization where single cells’ gene expression or cell-type labels are aggregated into equally sized pixels based on a user-defined resolution. SEraster is built on an R/Bioconductor S4 class called SpatialExperiment. SEraster can be incorporated with other packages to conduct downstream analyses for spatial omics datasets, such as detecting spatially variable genes.">
<meta property="og:description" content="SEraster is a rasterization preprocessing framework that aggregates cellular information into spatial pixels to reduce resource requirements for spatial omics data analysis. SEraster reduces the number of spatial points in spatial omics datasets for downstream analysis through a process of rasterization where single cells’ gene expression or cell-type labels are aggregated into equally sized pixels based on a user-defined resolution. SEraster is built on an R/Bioconductor S4 class called SpatialExperiment. SEraster can be incorporated with other packages to conduct downstream analyses for spatial omics datasets, such as detecting spatially variable genes.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SEraster</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><a class="dropdown-item" href="articles/formatting-SpatialExperiment-for-SEraster.html">Formatting a SpatialExperiment Object for SEraster</a></li>
    <li><a class="dropdown-item" href="articles/getting-started-with-SEraster.html">Getting Started With SEraster</a></li>
    <li><a class="dropdown-item" href="articles/SEraster-for-SVG-analysis.html">SEraster for Spatial Variable Genes Analysis</a></li>
    <li><a class="dropdown-item" href="articles/characterizing-mPOA-cell-type-heterogeneity.html">Characterizing mPOA cell-type heterogeneity with spatial bootstrapping</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JEFworks-Lab/SEraster/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="spatial-experiments-raster-seraster">Spatial Experiments raster (SEraster)<a class="anchor" aria-label="anchor" href="#spatial-experiments-raster-seraster"></a>
</h1></div>
<p><a href="https://github.com/JEFworks-Lab/SEraster/actions/workflows/check-standard.yaml" class="external-link"><img src="https://github.com/JEFworks-Lab/SEraster/actions/workflows/check-standard.yaml/badge.svg" alt="R-CMD-check"></a></p>
<p><code>SEraster</code> is a rasterization preprocessing framework that aggregates cellular information into spatial pixels to reduce resource requirements for spatial omics data analysis. This is the <code>SEraster</code> R documentation website. Questions, suggestions, or problems should be submitted as <a href="https://github.com/JEFworks-Lab/SEraster/issues" class="external-link">GitHub issues</a>.</p>
<p>
</p>
<p><img src="https://github.com/JEFworks/SEraster/blob/main/images/seraster_logo_hex.png?raw=true" align="center" height="300" style="float: center; height:300px;"></p>

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>SEraster</code> reduces the number of spatial points in spatial omics datasets for downstream analysis through a process of rasterization where single cells’ gene expression or cell-type labels are aggregated into equally sized pixels based on a user-defined <code>resolution</code>. Here, we refer to a particular <code>resolution</code> of rasterization by the side length of the pixel such that finer <code>resolution</code> indicates smaller pixel size and coarser <code>resolution</code> indicates larger pixel size.</p>
<p align="center">
</p>
<p><img src="https://github.com/JEFworks-Lab/SEraster/blob/main/images/overview.png?raw=true" height="600"></p>

</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install <code>SEraster</code>, we currently recommend using <code>remotes</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'JEFworks-Lab/SEraster'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<p>Introduction:</p>
<ul>
<li><a href="https://jef.works/SEraster/articles/formatting-SpatialExperiment-for-SEraster.html">Formatting a SpatialExperiment Object for SEraster</a></li>
<li><a href="https://jef.works/SEraster/articles/getting-started-with-SEraster.html">Getting Started With SEraster</a></li>
<li><a href="https://jef.works/SEraster/articles/SEraster-for-SVG-analysis.html">SEraster for Spatial Variable Genes Analysis</a></li>
<li><a href="https://jef.works/SEraster/articles/characterizing-mPOA-cell-type-heterogeneity.html">Characterizing mPOA cell-type heterogeneity with spatial bootstrapping</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Our manuscript describing <code>SEraster</code> is available on <em>Bioinformatics</em>:</p>
<p><a href="https://academic.oup.com/bioinformatics/article/40/7/btae412/7696710" class="external-link">Gohta Aihara, Kalen Clifton, Mayling Chen, Zhuoyan Li, Lyla Atta, Brendan F Miller, Rahul Satija, John W Hickey, Jean Fan, SEraster: a rasterization preprocessing framework for scalable spatial omics data analysis, Bioinformatics, Volume 40, Issue 7, July 2024, btae412, https://doi.org/10.1093/bioinformatics/btae412</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/JEFworks-Lab/SEraster/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/JEFworks-Lab/SEraster/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SEraster</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Gohta Aihara <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2492-9610" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Mayling Chen <br><small class="roles"> Author </small> <a href="https://orcid.org/0009-0009-0961-6665" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lyla Atta <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6113-0082" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jean Fan <br><small class="roles"> Author, reviewer </small> <a href="https://orcid.org/0000-0002-0212-5451" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gohta Aihara, Mayling Chen, Lyla Atta, Jean Fan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
