<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SEraster">
<title>Getting Started With SEraster • SEraster</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started With SEraster">
<meta property="og:description" content="SEraster">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SEraster</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Introduction</h6>
    <a class="dropdown-item" href="../articles/formatting-SpatialExperiment-for-SEraster.html">Formatting a SpatialExperiment Object for SEraster</a>
    <a class="dropdown-item" href="../articles/getting-started-with-SEraster.html">Getting Started With SEraster</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JEFworks-Lab/SEraster/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting Started With SEraster</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/JEFworks-Lab/SEraster/blob/HEAD/vignettes/getting-started-with-SEraster.Rmd" class="external-link"><code>vignettes/getting-started-with-SEraster.Rmd</code></a></small>
      <div class="d-none name"><code>getting-started-with-SEraster.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="getting-started-with-seraster">Getting Started with SEraster<a class="anchor" aria-label="anchor" href="#getting-started-with-seraster"></a>
</h2>
<p>This tutorial walks you through the basic functionalities of
<code>SEraster</code> and two examples of downstream analysis that can
be performed with the rasterized spatial omics data.</p>
<p>In the examples below, we assume the input data is provided as a
<code>SpatialExperiment</code> Bioconductor object. Please refer to the
following documentations to see how you would format your data into a
<code>SpatialExperiment</code> object:</p>
<ul>
<li>
<a href="https://bioconductor.org/packages/SpatialExperiment" class="external-link">SpatialExperiment</a>
package</li>
<li><a href="https://jef.works/SEraster/articles/formatting-SpatialExperiment-for-SEraster.html">Formatting
a SpatialExperiment Object for SEraster</a></li>
<li><a href="https://jef.works/SEraster/reference/merfish_mousePOA.html"><code>merfish_mousePOA</code>
dataset</a></li>
</ul>
<p>For downstream analyses, we will be using <a href="https://bioconductor.org/packages/nnSVG" class="external-link"><code>nnSVG</code></a>
for spatially variable gene (SVG) analysis and <a href="https://cran.r-project.org/package=CooccurrenceAffinity" class="external-link"><code>CooccurrenceAffinity</code></a>
for cell-type co-enrichment analysis.</p>
<p>References for nnSVG and CooccurrenceAffinity can be found below:</p>
<ul>
<li><a href="https://doi.org/10.1038/s41467-023-39748-z" class="external-link">Weber, L. et
al. (2023), “nnSVG for the scalable identification of spatially variable
genes using nearest-neighbor Gaussian processes”, <em>Nature
Communications</em></a></li>
<li><a href="https://doi.org/10.1126/sciadv.abj9204" class="external-link">Mainali, K. et
al. (2021), “A better index for analysis of co-occurrence and
similarity”, <em>Science Advances</em></a></li>
<li><a href="https://doi.org/10.1101/2022.11.01.514801" class="external-link">Mainali,K. et
al. (2022), “CooccurrenceAffinity: An R package for computing a novel
metric of affinity in co-occurrence data that corrects for pervasive
errors in traditional indices”, <em>bioRxiv</em></a></li>
</ul>
<div class="section level3">
<h3 id="load-libraries">Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/JEFworks-Lab/SEraster" class="external-link">SEraster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lmweber/nnSVG" class="external-link">nnSVG</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kpmainali/CooccurrenceAffinity" class="external-link">CooccurrenceAffinity</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: BiasedUrn</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-example-dataset">Load example dataset<a class="anchor" aria-label="anchor" href="#load-example-dataset"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"merfish_mousePOA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the dimension of the genes-by-cells matrix at single-cell resolution</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  155 6509</span></span>
<span></span>
<span><span class="co"># check the number of cell-types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 16</span></span></code></pre></div>
<p>This MERFISH mouse preoptic area (POA) dataset contains 6,509 cells
and 16 cell-types.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot at single-cell resolution</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span>, celltype <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, col <span class="op">=</span> <span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, stroke <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>col <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"x (μm)"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"y (μm)"</span>,</span>
<span>       col <span class="op">=</span> <span class="st">"Cell-types"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
</div>
<div class="section level3">
<h3 id="seraster-basic-functionalities">SEraster basic functionalities<a class="anchor" aria-label="anchor" href="#seraster-basic-functionalities"></a>
</h3>
<p><code>SEraster</code> reduces the number of spatial points in spatial
omics datasets for downstream analysis through a process of
rasterization where single cells’ gene expression or cell-type labels
are aggregated into equally sized square or hexagonal pixels (can be
changed using the <code>square</code> argument) based on a user-defined
resolution.</p>
<p>Here, we demonstrate the basic functionalities of
<code>SEraster</code>.</p>
<div class="section level4">
<h4 id="rasterize-gene-expression">Rasterize gene expression<a class="anchor" aria-label="anchor" href="#rasterize-gene-expression"></a>
</h4>
<p>For continuous variables such as gene expression or other molecular
information (e.g. protein expression if you are using spatial proteomics
datasets), <code>SEraster</code> aggregates the observed raw counts or
normalized expression values for each molecule within each pixel using
means by default (can be changed using the <code>fun</code>
argument).</p>
<p>Let’s try rasterizing the gene expression of the MERFISH mouse POA
dataset we loaded.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rastGexp</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the dimension of the genes-by-cells matrix after rasterizing gene expression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  155 1301</span></span></code></pre></div>
<p>As you can see, <code>SEraster</code> aggregated 6,509 single cells
into 1,301 pixels.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot total rasterized gene expression</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, name <span class="op">=</span> <span class="st">"Total rasterized gene expression"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot a specific gene</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, feature_name <span class="op">=</span> <span class="st">"Esr1"</span>, name <span class="op">=</span> <span class="st">"Esr1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-7-1.png" width="432"></p>
</div>
<div class="section level4">
<h4 id="rasterize-gene-expression-within-cell-type">Rasterize gene expression within cell-type<a class="anchor" aria-label="anchor" href="#rasterize-gene-expression-within-cell-type"></a>
</h4>
<p>Such rasterization can also be performed in a cell-type-specific
manner by restricting to cells of a particular cell-type prior to
rasterization. Here, we subset the dataset to Inhibitory cell-type and
run <code>SEraster</code> on the subsetted dataset.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## rasterize cell-type specific gene expression by subsetting to cell-type of interest</span></span>
<span><span class="va">ct_interest</span> <span class="op">&lt;-</span> <span class="st">"Inhibitory"</span></span>
<span><span class="va">spe_subset</span> <span class="op">&lt;-</span> <span class="va">merfish_mousePOA</span><span class="op">[</span>,<span class="va">merfish_mousePOA</span><span class="op">$</span><span class="va">celltype</span> <span class="op">==</span> <span class="va">ct_interest</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## rasterize gene expression</span></span>
<span><span class="va">rastGexpSubset</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">spe_subset</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## plot</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexpSubset</span>, name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Total rast gexp in "</span>, <span class="va">ct_interest</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-9-1.png" width="432"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexpSubset</span>, feature_name <span class="op">=</span> <span class="st">"Esr1"</span>, name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Esr1 in "</span>, <span class="va">ct_interest</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-10-1.png" width="432"></p>
</div>
<div class="section level4">
<h4 id="rasterize-cell-type">Rasterize cell-type<a class="anchor" aria-label="anchor" href="#rasterize-cell-type"></a>
</h4>
<p>For categorical variables such as cell-type or cluster labels,
<code>SEraster</code> aggregates the number of cells for each label
within each pixel using sums by default (can be changed using the
<code>fun</code> argument).</p>
<p>Let’s try rasterizing the cell-type labels of the MERFISH mouse POA
dataset.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rastCt</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeCellType.html">rasterizeCellType</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, col_name <span class="op">=</span> <span class="st">"celltype"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the dimension of the cell-types-by-cells matrix after rasterizing cell-type labels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  155 1301</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot total cell counts</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, name <span class="op">=</span> <span class="st">"cell counts"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-12-1.png" width="432"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot specific cell-type</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, feature_name <span class="op">=</span> <span class="st">"Inhibitory"</span>, name <span class="op">=</span> <span class="st">"Inhibitory neuron counts"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-13-1.png" width="432"></p>
</div>
<div class="section level4">
<h4 id="setting-rasterization-resolution">Setting rasterization resolution<a class="anchor" aria-label="anchor" href="#setting-rasterization-resolution"></a>
</h4>
<p>Rasterization resolution can be controlled by the
<code>resolution</code> argument of the
<code>rasterizeGeneExpression</code> and <code>rasterizeCellType</code>
functions. Here, we refer to a particular resolution of rasterization by
the side length for square pixels and the distance between opposite
edges for hexagonal pixels such that finer resolution indicates smaller
pixel size and vice versa.</p>
<p>Let’s see how the rasterized MERFISH mouse POA dataset look with
various resolutions using square pixels.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resolutions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">100</span>, <span class="fl">200</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">resolution</span> <span class="kw">in</span> <span class="va">resolutions</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># rasterize at defined resolution</span></span>
<span>  <span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="va">resolution</span><span class="op">)</span></span>
<span>  <span class="co"># plot a specific gene</span></span>
<span>  <span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">temp</span>, feature_name <span class="op">=</span> <span class="st">"Esr1"</span>, name <span class="op">=</span> <span class="st">"Esr1"</span>, plotTitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"resolution: "</span>, <span class="va">resolution</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">plt</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-14-1.png" width="432"><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-14-2.png" width="432"><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-14-3.png" width="432"></p>
<p>Now, let’s see the same resolutions using hexagonal pixels.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">resolution</span> <span class="kw">in</span> <span class="va">resolutions</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># rasterize at defined resolution</span></span>
<span>  <span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="va">resolution</span>, square <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="co"># plot a specific gene</span></span>
<span>  <span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">temp</span>, feature_name <span class="op">=</span> <span class="st">"Esr1"</span>, name <span class="op">=</span> <span class="st">"Esr1"</span>, plotTitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"resolution: "</span>, <span class="va">resolution</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">plt</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-15-1.png" width="432"><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-15-2.png" width="432"><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-15-3.png" width="432"></p>
</div>
<div class="section level4">
<h4 id="creating-and-rasterizing-permutations">Creating and rasterizing permutations<a class="anchor" aria-label="anchor" href="#creating-and-rasterizing-permutations"></a>
</h4>
<p>Since rasterized values may be sensitive to edge effects such as the
specific boundaries of grids upon rasterization, <code>SEraster</code>
enables permutation by rotating the dataset at various angles before
rasterization.</p>
<p>For example, let’s create 3 permutations of the MERFISH mouse POA
dataset, which would output a <code>list</code> of 3
<code>SpatialExperiment</code> objects with x,y coordinates rotated at
0, 120, and 240 degrees around the midrange point.</p>
<p>In addition to a single <code>SpatialExperiment</code> object,
<code>rasterizeGeneExpression</code> and <code>rasterizeCellType</code>
functions can both take a <code>list</code> of
<code>SpatialExperiment</code> objects. This essentially allows users to
streamline the preprocessing of permutations with <code>SEraster</code>;
followed by a downstream analysis of choice. For instance, in our
manuscript, we have shown that permutations can be used to improve the
performance of SVG analysis.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># permutate</span></span>
<span><span class="va">spe_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/permutateByRotation.html">permutateByRotation</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, n_perm <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rasterize permutated datasets at once</span></span>
<span><span class="va">out_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">spe_list</span>, assay_name <span class="op">=</span> <span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">out_list</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># extract rotated angle</span></span>
<span>  <span class="va">angle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"rotated_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rotated "</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out_list</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="st">" degrees"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># plot a specific gene</span></span>
<span>  <span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">out_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, feature_name <span class="op">=</span> <span class="st">"Esr1"</span>, name <span class="op">=</span> <span class="st">"Esr1"</span>, plotTitle <span class="op">=</span> <span class="va">angle</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">plt</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-16-1.png" width="432"><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-16-2.png" width="432"><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-16-3.png" width="432"></p>
<p>As you can see from the plots above, when <code>SEraster</code>
rasterizes a <code>list</code> of <code>SpatialExperiment</code>
objects, all <code>SpatialExperiment</code> objects in the inputted
<code>list</code> are rasterized with the same pixel coordinate
framework (same bounding box, resolution, centroid coordinates). This
feature may not be particularly useful for permutations; however, it can
potentially be applied to compare two or more datasets, such as
structurally aligned tissues as well as healthy vs. disease tissues.</p>
</div>
</div>
<div class="section level3">
<h3 id="examples-of-downstream-analyses-after-seraster-preprocessing">Examples of downstream analyses after SEraster preprocessing<a class="anchor" aria-label="anchor" href="#examples-of-downstream-analyses-after-seraster-preprocessing"></a>
</h3>
<div class="section level4">
<h4 id="spatial-variable-gene-svg-analysis">Spatial variable gene (SVG) analysis<a class="anchor" aria-label="anchor" href="#spatial-variable-gene-svg-analysis"></a>
</h4>
<p>Here, we use a previously developed tool called <code>nnSVG</code>.
Please refer to <a href="https://bioconductor.org/packages/nnSVG" class="external-link">nnSVG</a> for more
details about the package. We can directly input rasterized gene
expression <code>SpatialExperiment</code> object from
<code>SEraster</code> into <code>nnSVG</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run nnSVG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">rastGexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nnSVG/man/nnSVG.html" class="external-link">nnSVG</a></span><span class="op">(</span><span class="va">rastGexp</span>, assay_name <span class="op">=</span> <span class="st">"pixelval"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of significant SVGs based on the selected adjusted p-value threshold</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;    17   138</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot rasterized gene expression of top-ranked SVG</span></span>
<span><span class="va">top_svg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">$</span><span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">top_svg_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">top_svg</span><span class="op">]</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, feature_name <span class="op">=</span> <span class="va">top_svg_name</span>, name <span class="op">=</span> <span class="va">top_svg_name</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-19-1.png" width="432"></p>
<p>We can also perform cell-type specific SVG analysis by subsetting the
dataset prior to applying <code>SEraster</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset data</span></span>
<span><span class="va">ct_interest</span> <span class="op">&lt;-</span> <span class="st">"Excitatory"</span></span>
<span><span class="va">spe_sub</span> <span class="op">&lt;-</span> <span class="va">merfish_mousePOA</span><span class="op">[</span>,<span class="va">merfish_mousePOA</span><span class="op">$</span><span class="va">celltype</span> <span class="op">==</span> <span class="va">ct_interest</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># run SEraster</span></span>
<span><span class="va">rastGexp_sub</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">spe_sub</span>, assay_name<span class="op">=</span><span class="st">"volnorm"</span>, resolution <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run nnSVG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">rastGexp_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nnSVG/man/nnSVG.html" class="external-link">nnSVG</a></span><span class="op">(</span><span class="va">rastGexp_sub</span>, assay_name <span class="op">=</span> <span class="st">"pixelval"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of significant SVGs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp_sub</span><span class="op">)</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;    45   110</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot rasterized gene expression of top-ranked SVG</span></span>
<span><span class="va">top_svg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp_sub</span><span class="op">)</span><span class="op">$</span><span class="va">rank</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">top_svg_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">rastGexp_sub</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">top_svg</span><span class="op">]</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp_sub</span>, feature_name <span class="op">=</span> <span class="va">top_svg_name</span>, name <span class="op">=</span> <span class="va">top_svg_name</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-22-1.png" width="432"></p>
</div>
<div class="section level4">
<h4 id="cell-type-co-enrichment-analysis">Cell-type co-enrichment analysis<a class="anchor" aria-label="anchor" href="#cell-type-co-enrichment-analysis"></a>
</h4>
<p>Rasterized cell-type labels can be used to analyze pair-wise
cell-type co-enrichment To do so, we binarize the rasterized cell-type
labels using a relative enrichment metric and a previously developed
tool called <code>CooccurrenceAffinity</code>. Please refer to our paper
for more details about the methodology and <a href="https://CRAN.R-project.org/package=CooccurrenceAffinity" class="external-link">CooccurrenceAffinity</a>
for more details about the package.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract cell-type labels</span></span>
<span><span class="va">ct_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute relative enrichment (RE) metric</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">rastCt</span>, <span class="st">"pixelval"</span><span class="op">)</span></span>
<span><span class="va">mat_re</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">rastCt</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ct_label</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">mat</span><span class="op">[</span><span class="va">ct_label</span>,<span class="op">]</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mat</span><span class="op">[</span><span class="va">ct_label</span>,<span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat_re</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># binarize</span></span>
<span><span class="va">mat_bin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">mat_re</span> <span class="op">&gt;=</span> <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add RE and binarized layers to SpatialExperiment object</span></span>
<span><span class="fu">assays</span><span class="op">(</span><span class="va">rastCt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pixelval <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">rastCt</span>, <span class="st">"pixelval"</span><span class="op">)</span>, re <span class="op">=</span> <span class="va">mat_re</span>, bin <span class="op">=</span> <span class="va">mat_bin</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct_interest</span> <span class="op">&lt;-</span> <span class="st">"Ependymal"</span></span>
<span></span>
<span><span class="co"># plot pixel value for a cell-type of interest</span></span>
<span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, assay_name <span class="op">=</span> <span class="st">"pixelval"</span>, feature_name <span class="op">=</span> <span class="va">ct_interest</span>, name <span class="op">=</span> <span class="st">"cell-type counts"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-24-1.png" width="432"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot RE value for a cell-type of interest</span></span>
<span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, assay_name <span class="op">=</span> <span class="st">"re"</span>, feature_name <span class="op">=</span> <span class="va">ct_interest</span>, name <span class="op">=</span> <span class="st">"RE"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-25-1.png" width="432"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot binarized value for a cell-type of interest</span></span>
<span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastCt</span>, assay_name <span class="op">=</span> <span class="st">"bin"</span>, feature_name <span class="op">=</span> <span class="va">ct_interest</span>, factor_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"binarized"</span>, option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-26-1.png" width="432"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run CooccurrenceAffinity</span></span>
<span><span class="va">ct_coocc</span> <span class="op">&lt;-</span> <span class="fu">CooccurrenceAffinity</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/CooccurrenceAffinity/man/affinity.html" class="external-link">affinity</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mat_bin</span>, row.or.col <span class="op">=</span> <span class="st">"row"</span>, squarematrix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ------ as expected, the data ready for analysis has only 1 and 0... 1 = present, 0 = absent used for the interpretation ------</span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span><span class="co">#&gt; Warning in AcceptAffCI(x, marg, lev, Int2): NAs introduced by coercion</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AcceptAffCI(x, marg, lev, Int2): NAs introduced by coercion</span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = Infty is capped, along with upper confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = Infty is capped, along with upper confidence limits</span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span></span>
<span><span class="co">#&gt; Warning in AlphInts(x, marg, lev = lev, scal = scal, pvalType = pvalType): MLE = -Infty is capped, along with lower confidence limits</span></span>
<span><span class="co">#&gt; ~~~~~~~~~~ printing head of all elements of the output list ~~~~~~~~~~</span></span>
<span><span class="co">#&gt; $all</span></span>
<span><span class="co">#&gt;    entity_1      entity_2 entity_1_count_mA entity_2_count_mB obs_cooccur_X</span></span>
<span><span class="co">#&gt; 1 Ambiguous     Astrocyte               561               557           232</span></span>
<span><span class="co">#&gt; 2 Ambiguous Endothelial 1               561               308           136</span></span>
<span><span class="co">#&gt; 3 Ambiguous Endothelial 2               561                20             9</span></span>
<span><span class="co">#&gt; 4 Ambiguous Endothelial 3               561               102            49</span></span>
<span><span class="co">#&gt; 5 Ambiguous     Ependymal               561                47             4</span></span>
<span><span class="co">#&gt; 6 Ambiguous    Excitatory               561               504           192</span></span>
<span><span class="co">#&gt;   total_N    p_value exp_cooccur alpha_mle  alpha_medianInt conf_level</span></span>
<span><span class="co">#&gt; 1    1301     0.3657     240.182    -0.105 [-0.111, -0.098]       0.95</span></span>
<span><span class="co">#&gt; 2    1301      0.693     132.812     0.055   [0.047, 0.064]       0.95</span></span>
<span><span class="co">#&gt; 3    1301          1       8.624     0.077  [-0.021, 0.181]       0.95</span></span>
<span><span class="co">#&gt; 4    1301     0.2998      43.983     0.215   [0.194, 0.237]       0.95</span></span>
<span><span class="co">#&gt; 5    1301 2.1435e-07      20.267    -2.150 [-2.252, -1.986]       0.95</span></span>
<span><span class="co">#&gt; 6    1301     0.0041     217.328    -0.337  [-0.343, -0.33]       0.95</span></span>
<span><span class="co">#&gt;          ci_blaker            ci_cp          ci_midQ         ci_midP jaccard</span></span>
<span><span class="co">#&gt; 1  [-0.331, 0.119]  [-0.333, 0.123]  [-0.327, 0.117] [-0.327, 0.117]   0.262</span></span>
<span><span class="co">#&gt; 2  [-0.208, 0.313]  [-0.212, 0.321]  [-0.203, 0.312] [-0.203, 0.313]   0.186</span></span>
<span><span class="co">#&gt; 3   [-0.894, 1.04]  [-0.934, 1.061]  [-0.823, 0.959]  [-0.846, 0.98]   0.016</span></span>
<span><span class="co">#&gt; 4  [-0.201, 0.625]  [-0.212, 0.641]    [-0.19, 0.62] [-0.192, 0.622]   0.080</span></span>
<span><span class="co">#&gt; 5  [-3.27, -1.164]   [-3.5, -1.128] [-3.291, -1.223]  [-3.341, -1.2]   0.007</span></span>
<span><span class="co">#&gt; 6 [-0.569, -0.104] [-0.571, -0.104]  [-0.565, -0.11] [-0.565, -0.11]   0.220</span></span>
<span><span class="co">#&gt;   sorensen simpson errornote</span></span>
<span><span class="co">#&gt; 1    0.415   0.417        NA</span></span>
<span><span class="co">#&gt; 2    0.313   0.442        NA</span></span>
<span><span class="co">#&gt; 3    0.031   0.450        NA</span></span>
<span><span class="co">#&gt; 4    0.148   0.480        NA</span></span>
<span><span class="co">#&gt; 5    0.013   0.085        NA</span></span>
<span><span class="co">#&gt; 6    0.361   0.381        NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $occur_mat</span></span>
<span><span class="co">#&gt;         Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3 Ependymal</span></span>
<span><span class="co">#&gt; pixel23         1         0             0             0             0         0</span></span>
<span><span class="co">#&gt; pixel24         0         0             1             0             0         0</span></span>
<span><span class="co">#&gt; pixel25         0         1             0             0             0         0</span></span>
<span><span class="co">#&gt; pixel26         0         0             0             0             0         0</span></span>
<span><span class="co">#&gt; pixel27         1         0             1             0             0         0</span></span>
<span><span class="co">#&gt; pixel28         0         1             0             0             0         0</span></span>
<span><span class="co">#&gt;         Excitatory Inhibitory Microglia OD Immature 1 OD Immature 2 OD Mature 1</span></span>
<span><span class="co">#&gt; pixel23          0          1         0             0             0           0</span></span>
<span><span class="co">#&gt; pixel24          0          1         0             0             0           0</span></span>
<span><span class="co">#&gt; pixel25          0          0         0             1             0           0</span></span>
<span><span class="co">#&gt; pixel26          1          0         0             0             0           0</span></span>
<span><span class="co">#&gt; pixel27          0          0         0             0             0           0</span></span>
<span><span class="co">#&gt; pixel28          1          0         0             0             0           0</span></span>
<span><span class="co">#&gt;         OD Mature 2 OD Mature 3 OD Mature 4 Pericytes</span></span>
<span><span class="co">#&gt; pixel23           0           0           0         0</span></span>
<span><span class="co">#&gt; pixel24           0           0           0         0</span></span>
<span><span class="co">#&gt; pixel25           0           0           0         0</span></span>
<span><span class="co">#&gt; pixel26           0           0           0         0</span></span>
<span><span class="co">#&gt; pixel27           0           0           0         0</span></span>
<span><span class="co">#&gt; pixel28           0           0           0         0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $alpha_mle</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte        -0.105        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1     0.055    -0.015            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2     0.077    -0.118         0.779            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3     0.215    -0.029         0.322         1.902            NA</span></span>
<span><span class="co">#&gt; Ependymal        -2.150    -0.807        -0.143         0.345        -0.667</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $alpha_mle_sig</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA        NA            NA         1.902            NA</span></span>
<span><span class="co">#&gt; Ependymal         -2.15    -0.807            NA            NA            NA</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $p_value</span></span>
<span><span class="co">#&gt;                Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous           &lt;NA&gt;      &lt;NA&gt;          &lt;NA&gt;          &lt;NA&gt;            NA</span></span>
<span><span class="co">#&gt; Astrocyte         0.3657      &lt;NA&gt;          &lt;NA&gt;          &lt;NA&gt;            NA</span></span>
<span><span class="co">#&gt; Endothelial 1      0.693    0.9475          &lt;NA&gt;          &lt;NA&gt;            NA</span></span>
<span><span class="co">#&gt; Endothelial 2          1     0.825        0.1079          &lt;NA&gt;            NA</span></span>
<span><span class="co">#&gt; Endothelial 3     0.2998    0.9173        0.1813         5e-04            NA</span></span>
<span><span class="co">#&gt; Ependymal     2.1435e-07    0.0159        0.7323             1        0.4274</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous          &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte          &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1      &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2      &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3      &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal          &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $cooccur.null</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte       240.182        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1   132.812   131.865            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2     8.624     8.563         4.735            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3    43.983    43.669        24.148         1.568            NA</span></span>
<span><span class="co">#&gt; Ependymal        20.267    20.122        11.127         0.723         3.685</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $cooccur.obs</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte           232        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1       136       131            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2         9         8             8            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        49        43            30             7            NA</span></span>
<span><span class="co">#&gt; Ependymal             4        12            10             1             2</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $jaccard</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte         0.262        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1     0.186     0.178            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2     0.016     0.014         0.025            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3     0.080     0.070         0.079         0.061            NA</span></span>
<span><span class="co">#&gt; Ependymal         0.007     0.020         0.029         0.015         0.014</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $jaccard_sig</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA        NA            NA         0.061            NA</span></span>
<span><span class="co">#&gt; Ependymal         0.007      0.02            NA            NA            NA</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sorensen</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte         0.415        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1     0.313     0.303            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2     0.031     0.028         0.049            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3     0.148     0.131         0.146         0.115            NA</span></span>
<span><span class="co">#&gt; Ependymal         0.013     0.040         0.056         0.030         0.027</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sorensen_sig</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA        NA            NA         0.115            NA</span></span>
<span><span class="co">#&gt; Ependymal         0.013      0.04            NA            NA            NA</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $simpson</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte         0.417        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1     0.442     0.425            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2     0.450     0.400         0.400            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3     0.480     0.422         0.294          0.35            NA</span></span>
<span><span class="co">#&gt; Ependymal         0.085     0.255         0.213          0.05         0.043</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $simpson_sig</span></span>
<span><span class="co">#&gt;               Ambiguous Astrocyte Endothelial 1 Endothelial 2 Endothelial 3</span></span>
<span><span class="co">#&gt; Ambiguous            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA        NA            NA            NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA        NA            NA          0.35            NA</span></span>
<span><span class="co">#&gt; Ependymal         0.085     0.255            NA            NA            NA</span></span>
<span><span class="co">#&gt;               Ependymal Excitatory Inhibitory Microglia OD Immature 1</span></span>
<span><span class="co">#&gt; Ambiguous            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt; Ependymal            NA         NA         NA        NA            NA</span></span>
<span><span class="co">#&gt;               OD Immature 2 OD Mature 1 OD Mature 2 OD Mature 3 OD Mature 4</span></span>
<span><span class="co">#&gt; Ambiguous                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Astrocyte                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 1            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 2            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Endothelial 3            NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; Ependymal                NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt;               Pericytes</span></span>
<span><span class="co">#&gt; Ambiguous            NA</span></span>
<span><span class="co">#&gt; Astrocyte            NA</span></span>
<span><span class="co">#&gt; Endothelial 1        NA</span></span>
<span><span class="co">#&gt; Endothelial 2        NA</span></span>
<span><span class="co">#&gt; Endothelial 3        NA</span></span>
<span><span class="co">#&gt; Ependymal            NA</span></span>
<span><span class="co">#&gt; ~~~~~~~~~~ COMPLETED: printing head of all elements of the output list ~~~~~~~~~~</span></span>
<span></span>
<span><span class="co"># plot maximum likelihood estimates of affinity metric (alpha MLE)</span></span>
<span><span class="fu">CooccurrenceAffinity</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/CooccurrenceAffinity/man/plotgg.html" class="external-link">plotgg</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ct_coocc</span>, variable <span class="op">=</span> <span class="st">"alpha_mle"</span>, legendlimit <span class="op">=</span> <span class="st">"datarange"</span><span class="op">)</span></span>
<span><span class="co">#&gt; you can hide the printed values with show.value=F</span></span>
<span><span class="co">#&gt; use the argument value.digit to change number of digits and text.size to adjust the text size</span></span></code></pre></div>
<p><img src="getting-started-with-SEraster_files/figure-html/unnamed-chunk-27-1.png" width="432"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_3.5.1               CooccurrenceAffinity_1.0   </span></span>
<span><span class="co">#&gt;  [3] BiasedUrn_2.0.11            nnSVG_1.4.2                </span></span>
<span><span class="co">#&gt;  [5] SpatialExperiment_1.10.0    SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.52.1        GenomeInfoDb_1.36.4        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.34.1              S4Vectors_0.38.2           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.46.0         MatrixGenerics_1.12.3      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.3.0           SEraster_0.99.0            </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.2                 bitops_1.0-7             </span></span>
<span><span class="co">#&gt;   [3] pbapply_1.7-2             rlang_1.1.3              </span></span>
<span><span class="co">#&gt;   [5] magrittr_2.0.3            e1071_1.7-14             </span></span>
<span><span class="co">#&gt;   [7] compiler_4.3.1            DelayedMatrixStats_1.22.6</span></span>
<span><span class="co">#&gt;   [9] systemfonts_1.1.0         vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [11] pkgconfig_2.0.3           crayon_1.5.2             </span></span>
<span><span class="co">#&gt;  [13] rdist_0.0.5               fastmap_1.2.0            </span></span>
<span><span class="co">#&gt;  [15] backports_1.5.0           magick_2.8.3             </span></span>
<span><span class="co">#&gt;  [17] XVector_0.40.0            labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [19] scuttle_1.10.3            utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [21] BRISC_1.0.5               rmarkdown_2.27           </span></span>
<span><span class="co">#&gt;  [23] ragg_1.3.2                purrr_1.0.2              </span></span>
<span><span class="co">#&gt;  [25] xfun_0.44                 zlibbioc_1.46.0          </span></span>
<span><span class="co">#&gt;  [27] cachem_1.1.0              beachmat_2.16.0          </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.8.8            highr_0.11               </span></span>
<span><span class="co">#&gt;  [31] reshape_0.8.9             rhdf5filters_1.12.1      </span></span>
<span><span class="co">#&gt;  [33] DelayedArray_0.26.7       Rhdf5lib_1.22.1          </span></span>
<span><span class="co">#&gt;  [35] BiocParallel_1.36.0       parallel_4.3.1           </span></span>
<span><span class="co">#&gt;  [37] R6_2.5.1                  bslib_0.7.0              </span></span>
<span><span class="co">#&gt;  [39] limma_3.56.2              jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [41] Rcpp_1.0.12               knitr_1.45               </span></span>
<span><span class="co">#&gt;  [43] R.utils_2.12.3            Matrix_1.6-5             </span></span>
<span><span class="co">#&gt;  [45] tidyselect_1.2.1          rstudioapi_0.16.0        </span></span>
<span><span class="co">#&gt;  [47] abind_1.4-5               yaml_2.3.8               </span></span>
<span><span class="co">#&gt;  [49] codetools_0.2-20          plyr_1.8.9               </span></span>
<span><span class="co">#&gt;  [51] lattice_0.22-6            tibble_3.2.1             </span></span>
<span><span class="co">#&gt;  [53] withr_3.0.0               evaluate_0.23            </span></span>
<span><span class="co">#&gt;  [55] desc_1.4.3                sf_1.0-16                </span></span>
<span><span class="co">#&gt;  [57] units_0.8-5               proxy_0.4-27             </span></span>
<span><span class="co">#&gt;  [59] rearrr_0.3.4              pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [61] KernSmooth_2.23-24        checkmate_2.3.1          </span></span>
<span><span class="co">#&gt;  [63] generics_0.1.3            RCurl_1.98-1.14          </span></span>
<span><span class="co">#&gt;  [65] sparseMatrixStats_1.12.2  munsell_0.5.1            </span></span>
<span><span class="co">#&gt;  [67] scales_1.3.0              class_7.3-22             </span></span>
<span><span class="co">#&gt;  [69] glue_1.7.0                tools_4.3.1              </span></span>
<span><span class="co">#&gt;  [71] locfit_1.5-9.9            RANN_2.6.1               </span></span>
<span><span class="co">#&gt;  [73] fs_1.6.4                  cowplot_1.1.3            </span></span>
<span><span class="co">#&gt;  [75] rhdf5_2.44.0              grid_4.3.1               </span></span>
<span><span class="co">#&gt;  [77] DropletUtils_1.20.0       edgeR_3.42.4             </span></span>
<span><span class="co">#&gt;  [79] colorspace_2.1-0          GenomeInfoDbData_1.2.10  </span></span>
<span><span class="co">#&gt;  [81] HDF5Array_1.28.1          cli_3.6.2                </span></span>
<span><span class="co">#&gt;  [83] textshaping_0.4.0         fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [85] viridisLite_0.4.2         S4Arrays_1.0.6           </span></span>
<span><span class="co">#&gt;  [87] dplyr_1.1.4               gtable_0.3.5             </span></span>
<span><span class="co">#&gt;  [89] R.methodsS3_1.8.2         sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [91] digest_0.6.35             classInt_0.4-10          </span></span>
<span><span class="co">#&gt;  [93] dqrng_0.4.0               farver_2.1.2             </span></span>
<span><span class="co">#&gt;  [95] rjson_0.2.21              htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [97] memoise_2.0.1             htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [99] pkgdown_2.0.9             R.oo_1.26.0              </span></span>
<span><span class="co">#&gt; [101] lifecycle_1.0.4</span></span></code></pre></div>
</details>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gohta Aihara, Mayling Chen, Lyla Atta, Jean Fan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
