<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SEraster for Spatial Variable Genes Analysis • SEraster</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SEraster for Spatial Variable Genes Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SEraster</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><a class="dropdown-item" href="../articles/formatting-SpatialExperiment-for-SEraster.html">Formatting a SpatialExperiment Object for SEraster</a></li>
    <li><a class="dropdown-item" href="../articles/getting-started-with-SEraster.html">Getting Started With SEraster</a></li>
    <li><a class="dropdown-item" href="../articles/SEraster-for-SVG-analysis.html">SEraster for Spatial Variable Genes Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/characterizing-mPOA-cell-type-heterogeneity.html">Characterizing mPOA cell-type heterogeneity with spatial bootstrapping</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JEFworks-Lab/SEraster/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SEraster for Spatial Variable Genes Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/JEFworks-Lab/SEraster/blob/HEAD/vignettes/SEraster-for-SVG-analysis.Rmd" class="external-link"><code>vignettes/SEraster-for-SVG-analysis.Rmd</code></a></small>
      <div class="d-none name"><code>SEraster-for-SVG-analysis.Rmd</code></div>
    </div>

    
    
<p>In this tutorial, we will demonstrate how to use
<code>SEraster</code> as a preprocessing step prior to spatial variable
gene analysis with <code>MERINGUE</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/JEFworks-Lab/SEraster" class="external-link">SEraster</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="load-dataset">Load dataset<a class="anchor" aria-label="anchor" href="#load-dataset"></a>
</h2>
<p>We will use the MERFISH mPOA (bregma -0.29 slice from a female naive
animal) available in the SEraster package. The dataset is formatted as a
Spatial Experiment object as shown here: (<a href="https://jef.works/SEraster/reference/merfish_mousePOA.html" class="uri">https://jef.works/SEraster/reference/merfish_mousePOA.html</a>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"merfish_mousePOA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "SpatialExperiment"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "SpatialExperiment"</span></span></code></pre>
<p>Let’s first plot the total gene expression at the single-cell
resolution.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## pos is a matrix specifying a cell's x and y positions</span></span>
<span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span></span>
<span><span class="co">## counts is a genes-by-cells matrix where the gene expression was normalized by cell volume and scaled by 1000</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">assays</span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span><span class="op">$</span><span class="va">volnorm</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">pos</span>, total_gexp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">total_gexp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="rasterize-gene-expression-using-seraster">Rasterize gene expression using SEraster<a class="anchor" aria-label="anchor" href="#rasterize-gene-expression-using-seraster"></a>
</h2>
<p>The <code>rasterizeGeneExpression</code> function allows you to
rasterize gene expression at a resolution of your choosing. Below is an
example of rasterizing gene expression by calculating the mean
expression for all cells within each hexagonal pixel at 100 µm
resolution.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## let's rasterize the dataset at 100 um resolution.</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">rastGexp</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/rasterizeGeneExpression.html">rasterizeGeneExpression</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span>, </span>
<span>                                              assay_name <span class="op">=</span> <span class="st">"volnorm"</span>, </span>
<span>                                              resolution <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                              square <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                              fun <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span>
<span><span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, name <span class="op">=</span> <span class="st">"total_gexp"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p><code>SEraster</code> aggregated 6509 single-cells into 384 pixels,
with each pixel containing the normalized gene expression values using
mean. You can see how the dimension of <code>SpatialExperiment</code>
object is changed after rasterization.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">merfish_mousePOA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  155 6509</span></span>
<span><span class="co">## [1] 155 384</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="perform-svg-analysis-with-meringue">Perform SVG analysis with <code>MERINGUE</code><a class="anchor" aria-label="anchor" href="#perform-svg-analysis-with-meringue"></a>
</h2>
<p><code>MERINGUE</code> is a computational framework based on spatial
auto-correlation and cross-correlation analysis. It can be used to
identify genes with spatially heterogeneous expression patterns. Link to
install MERINGUE: (<a href="https://jef.works/MERINGUE/" class="external-link uri">https://jef.works/MERINGUE/</a>) First, let’s create an
adjacency weight matrix for the rasterized dataset.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">MERINGUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Get neighbor-relationships</span></span>
<span><span class="va">rastPos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu">MERINGUE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/getSpatialNeighbors.html" class="external-link">getSpatialNeighbors</a></span><span class="op">(</span><span class="va">rastPos</span>, filterDist <span class="op">=</span> <span class="fl">105</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/plotNetwork.html" class="external-link">plotNetwork</a></span><span class="op">(</span><span class="va">rastPos</span>, <span class="va">w</span><span class="op">)</span></span></code></pre></div>
<p><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>The adjacency weight matrix <code>w</code> contains the spatial
relationship of the 384 hexagonal pixels. Thus, each point in the plot
represents a hexagonal pixel. A line is drawn if the cells within that
pixel are spatial neighbors with cells within another pixels as defined
by the <code>filterDist</code> of 105. The <code>filterDist</code> is
the maximum Euclidean distance for which two cells cannot be considered
neighbors. This value was set based on the rasterized resolution of 100
um.</p>
<p>We can now calculate Moran’s I with the gene expression matrix and
identify significantly spatially auto-correlated genes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## rasterized genes-by-cells matrix, with cell pixels at 100 um resolution</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assays</span><span class="op">(</span><span class="va">rastGexp</span><span class="op">)</span><span class="op">$</span><span class="va">pixel</span></span>
<span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu">MERINGUE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/getSpatialPatterns.html" class="external-link">getSpatialPatterns</a></span><span class="op">(</span><span class="va">mat</span>, <span class="va">w</span><span class="op">)</span></span></code></pre></div>
<p>Based on Moran’s I calculations, 139 out of the 155 genes were
identified as significantly spatially auto-correlated genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results.filter</span> <span class="op">&lt;-</span> <span class="fu">MERINGUE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/filterSpatialPatterns.html" class="external-link">filterSpatialPatterns</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">mat</span>,</span>
<span>                                                  I <span class="op">=</span> <span class="va">I</span>,</span>
<span>                                                  w <span class="op">=</span> <span class="va">w</span>,</span>
<span>                                                  adjustPv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                  minPercentCells <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                  verbose <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                                  details <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Number of significantly autocorrelated genes: 139</span></span></code></pre>
<pre><code><span><span class="co">## ...driven by &gt; 19.2 cells: 128</span></span></code></pre>
<p>We can sort these spatially auto-correlated genes based their Moran’s
I statistic, as a higher I statistic generally correlates with greater
significance. Let’s identify the top three genes with the highest
Moran’s I statistics and visualize their expression patterns using the
<code>plotRaster</code> function in <code>SEraster</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## sort spatially auto-correlated genes with Moran's I statistic</span></span>
<span><span class="va">sortedResults</span> <span class="op">&lt;-</span> <span class="va">results.filter</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">results.filter</span><span class="op">$</span><span class="va">observed</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">## select 3 SVGs with the highest Moran's I</span></span>
<span><span class="va">sgenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sortedResults</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">sg</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sgenes</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co">## plot gene expression pattern</span></span>
<span>  <span class="va">plt</span> <span class="op">&lt;-</span> <span class="fu">SEraster</span><span class="fu">::</span><span class="fu"><a href="../reference/plotRaster.html">plotRaster</a></span><span class="op">(</span><span class="va">rastGexp</span>, </span>
<span>                              feature_name <span class="op">=</span> <span class="va">sg</span>,</span>
<span>                              name <span class="op">=</span> <span class="va">sg</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/show.html" class="external-link">show</a></span><span class="op">(</span><span class="va">plt</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-9-1.png" width="700"><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-9-2.png" width="700"><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-9-3.png" width="700"><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-9-4.png" width="700"></p>
<p>We also can compute the spatial cross correlation matrix to group
significantly spatially variable genes into primary spatial gene
expression patterns.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## create a duplicate filtered results with details = FALSE this time</span></span>
<span><span class="va">results.filter2</span> <span class="op">&lt;-</span> <span class="fu">MERINGUE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/filterSpatialPatterns.html" class="external-link">filterSpatialPatterns</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">mat</span>,</span>
<span>                                                  I <span class="op">=</span> <span class="va">I</span>,</span>
<span>                                                  w <span class="op">=</span> <span class="va">w</span>,</span>
<span>                                                  adjustPv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                  minPercentCells <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                                                  verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Number of significantly autocorrelated genes: 139</span></span></code></pre>
<pre><code><span><span class="co">## ...driven by &gt; 19.2 cells: 128</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## compute spatial cross correlation matrix</span></span>
<span><span class="va">scc</span> <span class="op">&lt;-</span> <span class="fu">MERINGUE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/spatialCrossCorMatrix.html" class="external-link">spatialCrossCorMatrix</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">[</span><span class="va">results.filter2</span>,<span class="op">]</span><span class="op">)</span>, </span>
<span>                                       weight <span class="op">=</span> <span class="va">w</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify primary patterns</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ggroup</span> <span class="op">&lt;-</span> <span class="fu">MERINGUE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MERINGUE/man/groupSigSpatialPatterns.html" class="external-link">groupSigSpatialPatterns</a></span><span class="op">(</span>pos <span class="op">=</span> <span class="va">rastPos</span>, </span>
<span>                                            mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">mat</span><span class="op">[</span><span class="va">results.filter2</span>,<span class="op">]</span><span class="op">)</span>, </span>
<span>                                            scc <span class="op">=</span> <span class="va">scc</span>, </span>
<span>                                            power <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                            hclustMethod <span class="op">=</span> <span class="st">'ward.D'</span>, </span>
<span>                                            deepSplit <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                            zlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Patterns detected:</span></span></code></pre>
<p><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-11-1.png" width="700"><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-11-2.png" width="700"><img src="SEraster-for-SVG-analysis_files/figure-html/unnamed-chunk-11-3.png" width="700"></p>
<pre><code><span><span class="co">##  ..cutHeight not given, setting it to 5.61  ===&gt;  99% of the (truncated) height range in dendro.</span></span>
<span><span class="co">##  ..done.</span></span>
<span><span class="co">## groups</span></span>
<span><span class="co">##  1  2  3  4  5  6  7  8  9 10 </span></span>
<span><span class="co">## 24 16 16 13 13 11 10 10  9  6</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gohta Aihara, Mayling Chen, Lyla Atta, Jean Fan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
